### ISSUE TITLE
[an report] pyminiaudio with Python3.12 on macos(COREAUDIO). 

### ISSUE THREAD

**Issue by noboruatkek** [irmen/pyminiaudio#75]
When installed pyminiaudio module using PIP, it did not work with Python3.12 on macos(14.5).
It just returns NULL devices for both playbacks, and captures.

Compiling  the pyminiaudio source code from Github with using the following command::

  env CFLAGS="-DMA_NO_RUNTIME_LINKING -framework CoreFoundation -framework CoreAudio -framework AudioToolbox " python3.12 setup.py clean build

solves my problem.  Although the options "-framework CoreFoundation -framework CoreAudio -framework AudioToolbox" resulted the warning messages,:

   clang: warning: -framework CoreFoundation: 'linker' input unused [-Wunused-command-line-argument]
   clang: warning: -framework CoreAudio: 'linker' input unused [-Wunused-command-line-argument]
   clang: warning: -framework AudioToolbox: 'linker' input unused [-Wunused-command-line-argument]

Thnaks,

**Comment by lanzz** [comment]
+1 I'm also running into the same issue, same setup (py3.12, macos 14.5).

**Comment by irmen** [comment]
Can you try again with the release 1.60 I've just uploaded?  I had to fix a few things to make the compiler happy, and it should also include mac os packages already.

**Comment by lanzz** [comment]
I'm still getting the null backend in the returned playback device and no actual playback.

**Comment by noboruatkek** [comment]
i still need to use: 

```
 env CFLAGS="-DMA_NO_RUNTIME_LINKING  -framework AudioToolbox " python3.12 setup.py clean build
```

to get a working binary library.

**Comment by noboruatkek** [comment]
The  "2.2. macOS and iOS" section ( miniaudio.h : lin 478) explains why we need to define CFLAGS as:
```
 env CFLAGS="-DMA_NO_RUNTIME_LINKING  -framework AudioUnit " python3.12 setup.py clean build
```

The following are  info on my test environment:
 Darwin Kernel Version 23.6.0: Sun Jun 30 19:42:22 PDT 2024; root:xnu-10063.140.33~20/RELEASE_ARM64_T8103 arm64 
 python 3.12.4

**Comment by irmen** [comment]
Ok so it seems this is related to the build process and not miniaudio.h itself.

**Comment by irmen** [comment]
@lanzz  @noboruatkek  can you tri again with the current master branch?  I tried adding the compiler options mentioned above to the mac os wheel builds.

**Comment by noboruatkek** [comment]
I cloned 1.61 from Github and run "make test".   Unfortunately it doesn't work. 

Compiler options used in  the  "make test"  process was:
```
clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -O3 -Wall -arch arm64 -arch x86_64 -g -DMA_NO_GENERATION=1 -DMA_NO_ENCODING=1 -DMA_NO_RESOURCE_MANAGER=1 -DMA_NO_NODE_GRAPH=1 -DMA_NO_ENGINE=1 -I/opt/src/pyminiaudio-1.6.1 -I/opt/src/pyminiaudio-1.6.1 -I/Library/Frameworks/Python.framework/Versions/3.12/include/python3.12 -c build/temp.macosx-10.9-universal2-cpython-312/_miniaudio.c -o build/temp.macosx-10.9-universal2-cpython-312/build/temp.macosx-10.9-universal2-cpython-312/_miniaudio.o -g1 -O3 -ffast-math
```
So there no "-DMA_NO_RUNTIME_LINKING" nor "framework AudioToolbox". 

By the way, I would greatly appreciate if you could take my suggestion on the distribution files.
1) In Makefiel: Use of "${PYTHON}" rather than "python". ( in my environment, "python" means "python2.7". I also hav both python3.9 and python3.12).
2) in test_devices:  check if playback devices and capture devices are not "NULL" devices.
I attached the files I used in the tests.
[Makefile.txt](https://github.com/user-attachments/files/16257545/Makefile.txt)
[test_miniaudio.txt](https://github.com/user-attachments/files/16257548/test_miniaudio.txt)

Thank you very much for your effort to distributing "pyminiaudio"!

**Comment by noboruatkek** [comment]
Pleas find a modified version of "build_ffi_module.py". It works for me.
[build_ffi_module.txt](https://github.com/user-attachments/files/16260660/build_ffi_module.txt)

**Comment by irmenrobidus** [comment]
Ah, I only changed the github action wheel builder config :|

**Comment by irmen** [comment]
I see I only modified the github action :|

However, since that did complete, there are pre-built packages available for download: https://github.com/irmen/pyminiaudio/actions/runs/9962348372   at the bottom is the zipfile with the build artifacts.

Can you perhaps confirm that the package that is in there also works correctly?  (this is what's getting uploaded to Pypi)

**Comment by noboruatkek** [comment]
1) I downloaded [pypi_artifacts](https://github.com/irmen/pyminiaudio/actions/runs/9962348372/artifacts/1707727957).
2) I setup .venv for test and activate it.
3) install miniaudio into this virtual env.
4) run python3.12  and got the following result.
```
 >>> devs=miniaudio.Devices()
>>> devs.get_playbacks()
[{'name': 'NULL Playback Device', 'type': <DeviceType.PLAYBACK: 1>, 'id': <cdata 'union ma_device_id *' owning 256 bytes>, 'formats': [{'format': 'Unknown', 'samplerate': 0, 'channels': 0}]}]
```

Working miniaudio module should return something like this:
```
>>> devs=miniaudio.Devices()
>>> devs.get_playbacks()
[{'name': 'MacBook Proのスピーカー', 'type': <DeviceType.PLAYBACK: 1>, 'id': <cdata 'union ma_device_id *' owning 256 bytes>, 'formats': [{'format': '32-bit IEEE Floating Point', 'samplerate': 44100, 'channels': 2}, {'format': '32-bit IEEE Floating Point', 'samplerate': 48000, 'channels': 2}, {'format': '32-bit IEEE Floating Point', 'samplerate': 88200, 'channels': 2}, {'format': '32-bit IEEE Floating Point', 'samplerate': 96000, 'channels': 2}]}, {'name': '複数出力装置', 'type': <DeviceType.PLAYBACK: 1>, 'id': <cdata 'union ma_device_id *' owning 256 bytes>, 'formats': [{'format': '32-bit IEEE Floating Point', 'samplerate': 44100, 'channels': 2}, {'format': '32-bit IEEE Floating Point', 'samplerate': 48000, 'channels': 2}, {'format': '32-bit IEEE Floating Point', 'samplerate': 88200, 'channels': 2}, {'format': '32-bit IEEE Floating Point', 'samplerate': 96000, 'channels': 2}]}]
```

**Comment by irmen** [comment]
I don't know how to fix the artifacts builds then :(

Because I believe I made the change already that should set those compiler options mentioned above.

I'm afraid we'll need help from someone more knowledgable with building wheels for MacOS

**Comment by irmen** [comment]
@noboruatkek  ok I modified the build_cffi file as well now, can you retry?  Also if you can, the latest build artefact?

**Comment by noboruatkek** [comment]
Both artifact: https://github.com/irmen/pyminiaudio/actions/runs/9997032405/artifacts/1716536082 and the latest source code works fine. 
(Please find logs for both artifact/build from source).
Thank you

**Comment by irmen** [comment]
I've published a new version 1.61 with the fixes!
